File: XCTestBootstrap/Reporters/FBJSONTestReporter.m
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBJSONTestReporter.h"

#import <XCTestBootstrap/XCTestBootstrap.h>

static inline NSString *FBFullyFormattedXCTestName(NSString *className, NSString *methodName);

@interface FBJSONTestReporter ()

@property (nonatomic, strong, readonly) id<FBDataConsumer> dataConsumer;
@property (nonatomic, strong, readonly) id<FBControlCoreLogger> logger;
@property (nonatomic, copy, readonly) NSString *testBundlePath;
@property (nonatomic, copy, readonly) NSString *testType;
@property (nonatomic, copy, readonly) NSMutableArray<NSDictionary<NSString *, id> *> *events;
@property (nonatomic, copy, readonly) NSMutableDictionary<NSString *, NSMutableArray<NSDictionary<NSString *, id> *> *> *xctestNameExceptionsMapping;
@property (nonatomic, copy, readonly) NSMutableArray<NSString *> *pendingTestOutput;

@property (nonatomic, copy, readwrite) NSString *currentTestName;
@property (nonatomic, assign, readwrite) BOOL started;
@property (nonatomic, assign, readwrite) BOOL finished;

@end

@implementation FBJSONTestReporter

- (instancetype)initWithTestBundlePath:(NSString *)testBundlePath testType:(NSString *)testType logger:(id<FBControlCoreLogger>)logger dataConsumer:(id<FBDataConsumer>)dataConsumer
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _dataConsumer = dataConsumer;
  _logger = logger;
  _testBundlePath = testBundlePath;
  _testType = testType;
  _xctestNameExceptionsMapping = [NSMutableDictionary dictionary];
  _pendingTestOutput = [NSMutableArray array];

  _currentTestName = nil;
  _started = NO;
  _finished = NO;

  return self;
}

- (BOOL)printReportWithError:(NSError **)error
{
  if (!_started) {
    return [[FBXCTestError describe:[self noStartOfTestPlanErrorMessage]] failBool:error];
  }
  if (!_finished) {
    NSString *errorMessage = @"No didFinishExecutingTestPlan event was received, the test bundle has likely crashed.";
    if (_currentTestName) {
      errorMessage = [errorMessage stringByAppendingString:@". Crash occurred while this test was running: "];
      errorMessage = [errorMessage stringByAppendingString:_currentTestName];
    }
    [self printEvent:[FBJSONTestReporter createOCUnitEndEvent:self.testType testBundlePath:self.testBundlePath message:errorMessage success:NO]];
    return [[FBXCTestError describe:errorMessage] failBool:error];
  }
  [self.dataConsumer consumeEndOfFile];
  return YES;
}

- (void)printEvent:(NSDictionary *)event
{
  NSMutableDictionary *timestamped = event.mutableCopy;
  if (!timestamped[@"timestamp"]) {
    timestamped[@"timestamp"] = @(NSDate.date.timeIntervalSince1970);
  }

  NSData *data = [NSJSONSerialization dataWithJSONObject:timestamped options:0 error:nil];
  [self.dataConsumer consumeData:data];
  [self.dataConsumer consumeData:[NSData dataWithBytes:"\n" length:1]];
}

#pragma mark FBXCTestReporter

- (void)processWaitingForDebuggerWithProcessIdentifier:(pid_t)pid
{
  [self printEvent:[FBJSONTestReporter waitingForDebuggerEvent:pid]];
}

- (void)debuggerAttached
{
  [self printEvent:[FBJSONTestReporter debuggerAttachedEvent]];
}

- (void)didBeginExecutingTestPlan
{
  _started = YES;
  [self printEvent:[FBJSONTestReporter createOCUnitBeginEvent:self.testType testBundlePath:self.testBundlePath]];
}

- (void)didFinishExecutingTestPlan
{
  if (_started) {
    [self printEvent:[FBJSONTestReporter createOCUnitEndEvent:self.testType testBundlePath:self.testBundlePath message:nil success:YES]];
  } else {
    [self printEvent:[FBJSONTestReporter createOCUnitBeginEvent:self.testType testBundlePath:self.testBundlePath]];
    NSString *errorMessage = [self noStartOfTestPlanErrorMessage];
    [self printEvent:[FBJSONTestReporter createOCUnitEndEvent:self.testType testBundlePath:self.testBundlePath message:errorMessage success:NO]];
  }
  _finished = YES;
}

- (void)testSuite:(NSString *)testSuite didStartAt:(NSString *)startTime
{
  [self printEvent:[FBJSONTestReporter beginTestSuiteEvent:testSuite]];
}

- (void)testCaseDidStartForTestClass:(NSString *)testClass method:(NSString *)method
{
  NSString *xctestName = FBFullyFormattedXCTestName(testClass, method);
  _currentTestName = xctestName;
  self.xctestNameExceptionsMapping[xctestName] = [NSMutableArray array];
  [self printEvent:[FBJSONTestReporter beginTestCaseEvent:testClass testMethod:method]];
}

- (void)testCaseDidFailForTestClass:(NSString *)testClass method:(NSString *)method withMessage:(NSString *)message file:(NSString *)file line:(NSUInteger)line
{
  NSString *xctestName = FBFullyFormattedXCTestName(testClass, method);
  [self.xctestNameExceptionsMapping[xctestName] addObject:[FBJSONTestReporter exceptionEvent:message file:file line:line]];
}

- (void)testCaseDidFinishForTestClass:(NSString *)testClass method:(NSString *)method withStatus:(FBTestReportStatus)status duration:(NSTimeInterval)duration logs:(NSArray<NSString *> *)logs
{
  _currentTestName = nil;
  NSDictionary<NSString *, id> *event = [FBJSONTestReporter
    testCaseDidFinishForTestClass:testClass
    method:method
    status:status
    duration:duration
    pendingTestOutput:self.pendingTestOutput
    xctestNameExceptionsMapping:self.xctestNameExceptionsMapping];
  [self printEvent:event];
  [self.pendingTestOutput removeAllObjects];
}

- (void)finishedWithSummary:(FBTestManagerResultSummary *)summary
{
  [self printEvent:[FBJSONTestReporter finishedEventFromSummary:summary]];
}

- (void)didRecordVideoAtPath:(nonnull NSString *)videoRecordingPath
{
  [self printEvent:@{
    @"event" : @"video-recording-finished",
    @"videoRecordingPath" : videoRecordingPath,
  }];
}

- (void)didSaveOSLogAtPath:(nonnull NSString *)osLogPath
{
  [self printEvent:@{
    @"event" : @"os-log-saved",
    @"osLogPath" : osLogPath,
  }];
}

- (void)didCopiedTestArtifact:(nonnull NSString *)testArtifactFilename toPath:(nonnull NSString *)path
{
  [self printEvent:@{
    @"event" : @"copy-test-artifact",
    @"test_artifact_file_name" : testArtifactFilename,
    @"path" : path,
  }];
}

- (void)testHadOutput:(NSString *)output
{
  [self.pendingTestOutput addObject:output];
  [self printEvent:[FBJSONTestReporter testOutputEvent:output]];
}

- (void)handleExternalEvent:(NSString *)line
{
  if (line.length == 0) {
    return;
  }
  NSError *error = nil;
  NSDictionary *event = [NSJSONSerialization JSONObjectWithData:[line dataUsingEncoding:NSUTF8StringEncoding] options:0 error:&error];
  if (event == nil) {
    [self.logger logFormat:@"Received unexpected output from otest-shim:\n%@", line];
  }
  if ([event[@"event"] isEqualToString:@"end-test"]) {
    NSMutableDictionary *mutableEvent = event.mutableCopy;
    mutableEvent[@"output"] = [self.pendingTestOutput componentsJoinedByString:@""];
    event = mutableEvent.copy;
    [self.pendingTestOutput removeAllObjects];
  }
  [self.events addObject:event];
}

- (void)appUnderTestExited
{
}


#pragma mark Event Synthesis

- (NSString *)noStartOfTestPlanErrorMessage
{
  NSString *errorMessage = @"No didBeginExecutingTestPlan event was received.";
  if (_currentTestName) {
    errorMessage = [errorMessage stringByAppendingString:@". However a test was running: "];
    errorMessage = [errorMessage stringByAppendingString:_currentTestName];
  }
  return errorMessage;
}

+ (NSDictionary<NSString *, id> *)exceptionEvent:(NSString *)reason file:(NSString *)file line:(NSUInteger)line
{
  return @{
    @"lineNumber" : @(line),
    @"filePathInProject" : file,
    @"reason" : reason,
  };
}

+ (NSDictionary<NSString *, id> *)beginTestCaseEvent:(NSString *)testClass testMethod:(NSString *)method
{
  return @{
    @"event" : @"begin-test",
    @"className" : testClass,
    @"methodName" : method,
    @"test" : FBFullyFormattedXCTestName(testClass, method),
  };
}

+ (NSDictionary<NSString *, id> *)beginTestSuiteEvent:(NSString *)testSuite
{
  return @{
    @"event" : @"begin-test-suite",
    @"suite" : testSuite,
  };
}

+ (NSDictionary<NSString *, id> *)testOutputEvent:(NSString *)output
{
  return @{
    @"event": @"test-output",
    @"output": output,
  };
}

+ (NSDictionary<NSString *, id> *)waitingForDebuggerEvent:(pid_t)pid
{
  return @{
    @"event": @"begin-status",
    @"pid": @(pid),
    @"level": @"Info",
    @"message": [NSString stringWithFormat:@"Tests waiting for debugger. To debug run: lldb -p %@", @(pid)],
  };
}

+ (NSDictionary<NSString *, id> *)debuggerAttachedEvent
{
  return @{
    @"event": @"end-status",
    @"level": @"Info",
    @"message": @"Debugger attached",
  };
}

+ (NSDictionary<NSString *, id> *)createOCUnitBeginEvent:(NSString *)testType testBundlePath:(NSString *)testBundlePath
{
  return @{
    @"event" : @"begin-ocunit",
    @"testType" : testType,
    @"bundleName" : [testBundlePath lastPathComponent],
    @"targetName" : testBundlePath,
  };
}

+ (NSDictionary<NSString *, id> *)createOCUnitEndEvent:(NSString *)testType testBundlePath:(NSString *)testBundlePath message:(NSString *)message success:(BOOL)success
{
  NSMutableDictionary<NSString *, id> *event = [NSMutableDictionary dictionaryWithDictionary:@{
    @"event" : @"end-ocunit",
    @"testType" : testType,
    @"bundleName" : [testBundlePath lastPathComponent],
    @"targetName" : testBundlePath,
    @"succeeded" : success ? @YES : @NO,
  }];
  if (message) {
    event[@"message"] = message;
  }
  return [event copy];
}

+ (NSDictionary<NSString *, id> *)finishedEventFromSummary:(FBTestManagerResultSummary *)summary
{
  return @{
    @"event" : @"end-test-suite",
    @"suite" : summary.testSuite,
    @"testCaseCount" : @(summary.runCount),
    @"totalFailureCount" : @(summary.failureCount),
    @"totalDuration" : @(summary.totalDuration),
    @"unexpectedExceptionCount" : @(summary.unexpected),
    @"testDuration" : @(summary.testDuration)
  };
}

+ (NSDictionary<NSString *, id> *)testCaseDidFinishForTestClass:(NSString *)testClass method:(NSString *)method status:(FBTestReportStatus)status duration:(NSTimeInterval)duration pendingTestOutput:(NSArray<NSString *> *)pendingTestOutput xctestNameExceptionsMapping:(NSDictionary<NSString *, NSArray<NSDictionary *> *> *)xctestNameExceptionsMapping
{
  NSString *xctestName = FBFullyFormattedXCTestName(testClass, method);
  return @{
    @"event" : @"end-test",
    @"result" : (status == FBTestReportStatusPassed ? @"success" : @"failure"),
    @"output" : [pendingTestOutput componentsJoinedByString:@""],
    @"test" : xctestName,
    @"className" : testClass,
    @"methodName" : method,
    @"succeeded" : (status == FBTestReportStatusPassed ? @YES : @NO),
    @"exceptions" : xctestNameExceptionsMapping[xctestName] ?: @[],
    @"totalDuration" : @(duration),
  };
}

@end

static inline NSString *FBFullyFormattedXCTestName(NSString *className, NSString *methodName)
{
  return [NSString stringWithFormat:@"-[%@ %@]", className, methodName];
}

File: XCTestBootstrap/Reporters/FBJSONTestReporter.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import FBControlCore
import Foundation

private func fullyFormattedXCTestName(_ className: String, _ methodName: String) -> String {
  return "-[\(className) \(methodName)]"
}

@objc public final class FBJSONTestReporter: NSObject, FBXCTestReporter {

  private let dataConsumer: FBDataConsumer
  private let logger: FBControlCoreLogger?
  private let testBundlePath: String
  private let testType: String
  private var events: [[String: Any]] = []
  private var xctestNameExceptionsMapping: [String: [[String: Any]]] = [:]
  private var pendingTestOutput: [String] = []

  private var currentTestName: String?
  private var crashError: Error?
  private var started: Bool = false
  private var finished: Bool = false

  @objc public init(testBundlePath: String, testType: String, logger: FBControlCoreLogger?, dataConsumer: FBDataConsumer) {
    self.dataConsumer = dataConsumer
    self.logger = logger
    self.testBundlePath = testBundlePath
    self.testType = testType
    super.init()
  }

  // MARK: FBXCTestReporter

  @objc public func printReport() throws {
    if !started {
      throw XCTestBootstrapError.describe(noStartOfTestPlanErrorMessage()).build()
    }
    if !finished {
      var errorMessage = "No didFinishExecutingTestPlan event was received, the test bundle has likely crashed."
      if let crashError {
        errorMessage = crashError.localizedDescription
      }
      if let currentTestName {
        errorMessage += ". Crash occurred while this test was running: \(currentTestName)"
      }
      printEvent(FBJSONTestReporter.createOCUnitEndEvent(testType, testBundlePath: testBundlePath, message: errorMessage, success: false))
      throw XCTestBootstrapError.describe(errorMessage).build()
    }
    dataConsumer.consumeEndOfFile()
  }

  @objc public func processWaitingForDebugger(withProcessIdentifier pid: pid_t) {
    printEvent(FBJSONTestReporter.waitingForDebuggerEvent(pid))
  }

  @objc public func didBeginExecutingTestPlan() {
    started = true
    printEvent(FBJSONTestReporter.createOCUnitBeginEvent(testType, testBundlePath: testBundlePath))
  }

  @objc public func didFinishExecutingTestPlan() {
    if started {
      printEvent(FBJSONTestReporter.createOCUnitEndEvent(testType, testBundlePath: testBundlePath, message: nil, success: true))
    } else {
      printEvent(FBJSONTestReporter.createOCUnitBeginEvent(testType, testBundlePath: testBundlePath))
      let errorMessage = noStartOfTestPlanErrorMessage()
      printEvent(FBJSONTestReporter.createOCUnitEndEvent(testType, testBundlePath: testBundlePath, message: errorMessage, success: false))
    }
    finished = true
  }

  @objc public func testSuite(_ testSuite: String, didStartAt startTime: String) {
    printEvent(FBJSONTestReporter.beginTestSuiteEvent(testSuite))
  }

  @objc public func testCaseDidStart(forTestClass testClass: String, method: String) {
    let xctestName = fullyFormattedXCTestName(testClass, method)
    currentTestName = xctestName
    xctestNameExceptionsMapping[xctestName] = []
    printEvent(FBJSONTestReporter.beginTestCaseEvent(testClass, testMethod: method))
  }

  @objc public func testCaseDidFail(forTestClass testClass: String, method: String, exceptions: [FBExceptionInfo]) {
    let xctestName = fullyFormattedXCTestName(testClass, method)
    for exception in exceptions {
      xctestNameExceptionsMapping[xctestName]?.append(FBJSONTestReporter.exceptionEvent(exception.message, file: exception.file ?? "", line: exception.line))
    }
  }

  @objc public func testCaseDidFinish(forTestClass testClass: String, method: String, with status: FBTestReportStatus, duration: TimeInterval, logs: [String]?) {
    currentTestName = nil
    let event = FBJSONTestReporter.testCaseDidFinishEvent(forTestClass: testClass, method: method, status: status, duration: duration, pendingTestOutput: pendingTestOutput, xctestNameExceptionsMapping: xctestNameExceptionsMapping)
    printEvent(event)
    pendingTestOutput.removeAll()
  }

  @objc public func finished(with summary: FBTestManagerResultSummary) {
    printEvent(FBJSONTestReporter.finishedEvent(from: summary))
  }

  @objc public func didRecordVideo(atPath videoRecordingPath: String) {
    printEvent([
      "event": "video-recording-finished",
      "videoRecordingPath": videoRecordingPath,
    ])
  }

  @objc public func didSaveOSLog(atPath osLogPath: String) {
    printEvent([
      "event": "os-log-saved",
      "osLogPath": osLogPath,
    ])
  }

  @objc public func didCopiedTestArtifact(_ testArtifactFilename: String, toPath path: String) {
    printEvent([
      "event": "copy-test-artifact",
      "test_artifact_file_name": testArtifactFilename,
      "path": path,
    ])
  }

  @objc public func testHadOutput(_ output: String) {
    pendingTestOutput.append(output)
    printEvent(FBJSONTestReporter.testOutputEvent(output))
  }

  @objc public func handleExternalEvent(_ line: String) {
    if line.isEmpty {
      return
    }
    guard let data = line.data(using: .utf8),
      var event = try? JSONSerialization.jsonObject(with: data, options: []) as? [String: Any]
    else {
      logger?.log("Received unexpected output from otest-shim:\n\(line)")
      return
    }
    if event["event"] as? String == "end-test" {
      event["output"] = pendingTestOutput.joined()
      pendingTestOutput.removeAll()
    }
    events.append(event)
  }

  @objc public func processUnderTestDidExit() {}

  @objc public func didCrashDuringTest(_ error: Error) {
    crashError = error
  }

  // MARK: Private

  private func printEvent(_ event: [String: Any]) {
    var timestamped = event
    if timestamped["timestamp"] == nil {
      timestamped["timestamp"] = Date().timeIntervalSince1970
    }
    if let data = try? JSONSerialization.data(withJSONObject: timestamped, options: []) {
      dataConsumer.consumeData(data)
      dataConsumer.consumeData(Data("\n".utf8))
    }
  }

  private func noStartOfTestPlanErrorMessage() -> String {
    var errorMessage = "No didBeginExecutingTestPlan event was received."
    if let currentTestName {
      errorMessage += ". However a test was running: \(currentTestName)"
    }
    return errorMessage
  }

  private static func exceptionEvent(_ reason: String, file: String, line: UInt) -> [String: Any] {
    return [
      "lineNumber": line,
      "filePathInProject": file,
      "reason": reason,
    ]
  }

  private static func beginTestCaseEvent(_ testClass: String, testMethod method: String) -> [String: Any] {
    return [
      "event": "begin-test",
      "className": testClass,
      "methodName": method,
      "test": fullyFormattedXCTestName(testClass, method),
    ]
  }

  private static func beginTestSuiteEvent(_ testSuite: String) -> [String: Any] {
    return [
      "event": "begin-test-suite",
      "suite": testSuite,
    ]
  }

  private static func testOutputEvent(_ output: String) -> [String: Any] {
    return [
      "event": "test-output",
      "output": output,
    ]
  }

  private static func waitingForDebuggerEvent(_ pid: pid_t) -> [String: Any] {
    return [
      "event": "begin-status",
      "pid": pid,
      "level": "Info",
      "message": "Tests waiting for debugger. To debug run: lldb -p \(pid)",
    ]
  }

  private static func createOCUnitBeginEvent(_ testType: String, testBundlePath: String) -> [String: Any] {
    return [
      "event": "begin-ocunit",
      "testType": testType,
      "bundleName": (testBundlePath as NSString).lastPathComponent,
      "targetName": testBundlePath,
    ]
  }

  private static func createOCUnitEndEvent(_ testType: String, testBundlePath: String, message: String?, success: Bool) -> [String: Any] {
    var event: [String: Any] = [
      "event": "end-ocunit",
      "testType": testType,
      "bundleName": (testBundlePath as NSString).lastPathComponent,
      "targetName": testBundlePath,
      "succeeded": success,
    ]
    if let message {
      event["message"] = message
    }
    return event
  }

  private static func finishedEvent(from summary: FBTestManagerResultSummary) -> [String: Any] {
    return [
      "event": "end-test-suite",
      "suite": summary.testSuite,
      "testCaseCount": summary.runCount,
      "totalFailureCount": summary.failureCount,
      "totalDuration": summary.totalDuration,
      "unexpectedExceptionCount": summary.unexpected,
      "testDuration": summary.testDuration,
    ]
  }

  private static func testCaseDidFinishEvent(forTestClass testClass: String, method: String, status: FBTestReportStatus, duration: TimeInterval, pendingTestOutput: [String], xctestNameExceptionsMapping: [String: [[String: Any]]]) -> [String: Any] {
    let xctestName = fullyFormattedXCTestName(testClass, method)
    return [
      "event": "end-test",
      "result": (status == .passed ? "success" : "failure"),
      "output": pendingTestOutput.joined(),
      "test": xctestName,
      "className": testClass,
      "methodName": method,
      "succeeded": (status == .passed),
      "exceptions": xctestNameExceptionsMapping[xctestName] ?? [],
      "totalDuration": duration,
    ]
  }
}
